### PR TITLE
*Finish registration functionality

### PR DESCRIPTION
**Pull request by maxonchil**
*Added abbility to give out a token, when user is registrated(save it to
LS)
*Added password hashing
*Added redirect to personal page, if login successful(without rendering)
*Replaced middlewar to fix cross-origin issue for module 'cros'
*Change a User schema(added requirments and validation)
*Added errors handle(when some form fields are clear or user with such
nickname registrer(last one improved))
*Added config variables
*Suffered regexp's in angular to env variables

### CODE REVIEW

File: homework3/semi-uber/Server/api/Schemas/user.schema.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+const mongoose = require("mongoose");
+const uniqueValidator = require("mongoose-unique-validator");
+const Schema = mongoose.Schema;
+const User = new Schema({
+  name: { type: String, require: true, minLength: [3, "Name is too short!"] },
```
Comment (MaxHouz) [inline homework3/semi-uber/Server/api/Schemas/user.schema.js]: and what about Li, Po ? :) They will not be able to work in your system?)

File: homework3/semi-uber/Server/index.js [hunk 2]
```diff
@@ -14,8 +14,10 @@ mongoose.connect(dbURL, {
   useNewUrlParser: true
 });
 
+app.use(cors());
+
 app.use(express.json());
-app.use(allowCrossDomain);
+
 app.use("../api", express.static("api"));
```
Comment (MaxHouz) [inline homework3/semi-uber/Server/index.js]: still have this quetsion. Will you use some static files in your application?

File: homework3/semi-uber/UI/src/app/registration-page/registration-page.component.html [hunk 3]
```diff
@@ -90,9 +90,11 @@
...
     </label>
 
-    <button class="registration__button button" (click)="getRegForm()">
+    <button class="registration__button button" (click)="registrateUser()">
       Submit
     </button>
   </div>
 </form>
-<span class="registration__error" *ngIf="formField">All fields must be filled</span>
+<span class="registration__error" *ngIf="errorFillFields"
```
Comment (MaxHouz) [inline homework3/semi-uber/UI/src/app/registration-page/registration-page.component.html]: I've already written you about the validation. As for me it is better to show error messages under every invalid input with the corresponding error. For example if user started typing email, but it is not valid and user became typing in the other input, you should display `Invalid email`. So every input that has some special validation rules should display the corresponding error for every concrete validation rule, that is not valid

File: homework3/semi-uber/UI/src/app/registration-page/registration-page.component.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      name,
+      username,
+      password,
+      email,
+      status
+    };
+    this.http
+      .post(env.regUrl, this.user, { observe: "response" })
+      .subscribe((res: HttpResponse<any>) => console.log(res));
+    this.router.navigate(["/user/:1"]);
```
Comment (MaxHouz) [inline homework3/semi-uber/UI/src/app/registration-page/registration-page.component.ts]: you should do navigation only after the request was done

File: homework3/semi-uber/UI/src/environments/environment.ts [hunk 5]
```diff
@@ -4,7 +4,10 @@
 
 export const environment = {
   production: false,
-  regUrl: "http://localhost:3000/registration"
+  regUrl: "http://localhost:3000/registration",
+  pasRegexp:
```
Comment (MaxHouz) [inline homework3/semi-uber/UI/src/environments/environment.ts]: Bad idea. regular expressions are not dependent on environment, so there is no reason to put them to env variables

File: homework3/semi-uber/UI/src/environments/environment.ts [hunk 6]
```diff
@@ -4,7 +4,10 @@
...
+    "(?=^.{6,}$)((?=.*d)(?=.*[A-Z])(?=.*[a-z])|(?=.*d)(?=.*[^A-Za-z0-9])(?=.*[a-z])|(?=.*[^A-Za-z0-9])(?=.*[A-Z])(?=.*[a-z])|(?=.*d)(?=.*[A-Z])(?=.*[^A-Za-z0-9]))^.*",
+  logRegexp: ".{5,10}"
```
Comment (MaxHouz) [inline homework3/semi-uber/UI/src/environments/environment.ts]: same

File: homework3/semi-uber/Server/api/routes/registration.router.js [hunk 7]
```diff
@@ -1,40 +1,47 @@
...
-  User.find({ username }).then(answer =>
-    answer.length === 0
-      ? saveUser(user)
-      : res.json({ status: "Nickname is allready taken" })
-  );
+  hashPassword(password, saltRounds)
+    .then(password => createToken({ username, password }, secret))
+    .then(
+      res =>
+        new User({
```
Comment (MaxHouz) [inline homework3/semi-uber/Server/api/routes/registration.router.js]: I would suggest you to start using `Joi` already, now is the best moment to add it to your project.

File: homework3/semi-uber/Server/index.js (hunk 2 above)
Comment (maxonchil) [reply homework3/semi-uber/Server/index.js]: At that moment - no

File: homework3/semi-uber/UI/src/environments/environment.ts (hunk 5 above)
Comment (maxonchil) [reply homework3/semi-uber/UI/src/environments/environment.ts]: fixed

File: homework3/semi-uber/UI/src/environments/environment.ts (hunk 6 above)
Comment (maxonchil) [reply homework3/semi-uber/UI/src/environments/environment.ts]: fixed

File: homework3/semi-uber/Server/api/Schemas/user.schema.js (hunk 1 above)
Comment (maxonchil) [reply homework3/semi-uber/Server/api/Schemas/user.schema.js]: Fixed

File: homework3/semi-uber/UI/src/app/registration-page/registration-page.component.ts (hunk 4 above)
Comment (maxonchil) [reply homework3/semi-uber/UI/src/app/registration-page/registration-page.component.ts]: Fixed

File: homework3/semi-uber/Server/api/routes/registration.router.js (hunk 7 above)
Comment (maxonchil) [reply homework3/semi-uber/Server/api/routes/registration.router.js]: Done

File: homework3/semi-uber/UI/src/app/registration-page/registration-page.component.html (hunk 3 above)
Comment (maxonchil) [reply homework3/semi-uber/UI/src/app/registration-page/registration-page.component.html]: Notice it for future

**Pull request merged**